### PR TITLE
[2.13.x] DDF-4472 Replacing the DescribableServiceMap with Java8 Streams /filters

### PR DESCRIPTION
**Pull request by Kjames5269**
#### What does this PR do?
Fixes a bug where a Concurrency Control Exception was occurring while polling sources by removing the Describable Service Map

#### Who is reviewing it? 
@mcalcote @bakejeyner 

#### Select relevant component teams: 
@codice/core-apis 

#### Ask 2 committers to review/merge the PR and tag them here.

@brendan-hofmann
@emmberk 

#### How should this be tested?

Add a source to DDF, query and ensure you get results without errors.

#### What are the relevant tickets?
[DDF-4472](https://codice.atlassian.net/browse/DDF-4472)
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/core/catalog-core-standardframework/src/test/groovy/ddf/catalog/impl/operations/ResourceOperationsSpec.groovy [hunk 1]
```diff
@@ -192,4 +192,11 @@ class ResourceOperationsSpec extends Specification {
         policyResponse.operationPolicy() >> [:]
         return policyResponse
     }
+
+    FederatedSource mockFederatedSource(def id) {
```
Comment (emmberk) [inline catalog/core/catalog-core-standardframework/src/test/groovy/ddf/catalog/impl/operations/ResourceOperationsSpec.groovy]: ✏️ These mocks can be inlined instead. I don't think this helper method provides much value:
```groovy
federatedSources = [Mock(FederatedSource) {
    getId() >> 'resourcename'
}]
```

File: catalog/core/catalog-core-standardframework/src/test/groovy/ddf/catalog/impl/operations/SourceOperationsSpec.groovy [hunk 2]
```diff
@@ -498,4 +501,8 @@ class SourceOperationsSpec extends Specification {
 
         return fedSource
     }
+
+    private static List<FederatedSource> getKeyList(def fedSources) {
+        return fedSources.stream().map({ source -> source.getId() }).collect(Collectors.toList())
```
Comment (emmberk) [inline catalog/core/catalog-core-standardframework/src/test/groovy/ddf/catalog/impl/operations/SourceOperationsSpec.groovy]: ✏️ Could be groovier and inlined:
```groovy
fedSources.collect { it.id }
```

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/SourceOperations.java [hunk 3]
```diff
@@ -201,22 +207,29 @@ public SourceInfoResponse getSourceInfo(
...
-            LOGGER.debug("Found federated source: {}", requestedSourceId);
-            discoveredSources.add(frameworkProperties.getFederatedSources().get(requestedSourceId));
-          }
-          if (!containsId) {
+          final Optional<FederatedSource> source =
+              frameworkProperties
+                  .getFederatedSources()
+                  .stream()
+                  .filter(e -> e.getId().equals(requestedSourceId))
+                  .findFirst();
```
Comment (emmberk) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/SourceOperations.java]: ❓ Is there a risk of there being more than one source with the `requestedSourceId`? Should this case be handled (e.g. log something)?

Comment (Kjames5269) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/SourceOperations.java]: Good question, I originally had a list and was always getting the first element and thought this looked cleaner. If the ID isn't a unique key then this is definitely a possibility that could happen but I'm not entirely sure how we'd want to handle that.

**Comment by Kjames5269** [comment]
build now

**Comment by Kjames5269** [comment]
build now

**Comment by Kjames5269** [comment]
build now

**Comment by Kjames5269** [comment]
build now

**Comment by Kjames5269** [comment]
build now

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java [hunk 4]
```diff
@@ -215,11 +215,11 @@ public void setConnectedSources(List<ConnectedSource> connectedSources) {
     this.connectedSources = connectedSources;
   }
 
-  public Map<String, FederatedSource> getFederatedSources() {
+  public Collection<FederatedSource> getFederatedSources() {
     return federatedSources;
   }
 
-  public void setFederatedSources(Map<String, FederatedSource> federatedSources) {
+  public void setFederatedSources(Collection<FederatedSource> federatedSources) {
```
Comment (brendan-hofmann) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: ❓ Should we make a defensive copy here?

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java [hunk 5]
```diff
@@ -281,12 +281,12 @@ public void setReliableResourceDownloadManager(
     this.reliableResourceDownloadManager = reliableResourceDownloadManager;
   }
 
-  public Map<String, CatalogStore> getCatalogStoresMap() {
-    return catalogStoresMap;
+  public Collection<CatalogStore> getCatalogStores() {
+    return catalogStores;
```
Comment (brendan-hofmann) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: ❓ Should we return an (unmodifiable) defensive copy here?

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java [hunk 6]
```diff
@@ -57,7 +57,7 @@
 
   private List<CatalogProvider> catalogProviders = new ArrayList<>();
 
-  private Map<String, CatalogStore> catalogStoresMap = new HashMap<>();
+  private Collection<CatalogStore> catalogStores = Collections.emptyList();
```
Comment (brendan-hofmann) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: ❓ Should we use a thread-safe collection here?

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java [hunk 7]
```diff
@@ -81,7 +81,7 @@
 
   private List<ConnectedSource> connectedSources = new ArrayList<>();
 
-  private Map<String, FederatedSource> federatedSources = new HashMap<>();
+  private Collection<FederatedSource> federatedSources = Collections.emptyList();
```
Comment (brendan-hofmann) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: ❓ Should we use a thread-safe collection here?

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java [hunk 8]
```diff
@@ -281,12 +281,12 @@ public void setReliableResourceDownloadManager(
...
-  public Map<String, CatalogStore> getCatalogStoresMap() {
-    return catalogStoresMap;
+  public Collection<CatalogStore> getCatalogStores() {
+    return catalogStores;
   }
 
-  public void setCatalogStoresMap(Map<String, CatalogStore> catalogStoresMap) {
-    this.catalogStoresMap = catalogStoresMap;
+  public void setCatalogStores(Collection<CatalogStore> catalogStores) {
+    this.catalogStores = catalogStores;
```
Comment (brendan-hofmann) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: ❓ Defensive copy?

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/SourceOperations.java (hunk 3 above)
Comment (brendan-hofmann) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/SourceOperations.java]: It is expected but not guaranteed/enforced to be unique.

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java (hunk 4 above)
Comment (Kjames5269) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: Shouldn't need to, blueprint should inject this as a [Dynamic Collection](https://github.com/apache/aries/blob/trunk/blueprint/blueprint-core/src/main/java/org/apache/aries/blueprint/utils/DynamicCollection.java) that takes care of concurrency control for us.

Comment (Kjames5269) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: As for making it non modifiable that's a good question. I don't think we'd need to as nothing else in the framework properties does that

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java (hunk 6 above)
Comment (Kjames5269) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: see [above](https://github.com/codice/ddf/pull/4250#discussion_r256468534)

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java (hunk 7 above)
Comment (Kjames5269) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: see [above](https://github.com/codice/ddf/pull/4250#discussion_r256467716)

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java (hunk 8 above)
Comment (Kjames5269) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/FrameworkProperties.java]: see [above](https://github.com/codice/ddf/pull/4250#discussion_r256468873)

**Comment by Kjames5269** [comment]
build now

**Comment by Kjames5269** [comment]
build now

**Comment by vinamartin** [comment]
✅ Manually verified after rebase. LGTM

**Comment by Kjames5269** [comment]
build now

**Comment by Kjames5269** [comment]
build now

**Comment by vinamartin** [comment]
Overriding Codacy/PR Quality Review check which seemed to get stuck

**Pull request merged**